### PR TITLE
Log warn and error from terraform run

### PR DESCRIPTION
**Pull request by hemslo**
Log `[WARN]` and `[ERROR]` from terraform output, so we can get more details from unexpected errors.

[APPSRE-7884](https://issues.redhat.com/browse/APPSRE-7884)

### <samp>🤖 Generated by Copilot at f9f4e97</samp>

Added logging features and tests for `TerraformClient` class. This improves the visibility and error handling of the terraform operations performed by the class and its methods. The logging captures both the terraform log output and the provider logs. The tests cover the new `check_output` method that parses the log output.

### <samp>🤖 Generated by Copilot at f9f4e97</samp>

*  Add context manager to capture and handle terraform log output in `terraform_init`, `terraform_output`, `terraform_plan`, and `terraform_apply` methods of `TerraformClient` class ([link](https://github.com/app-sre/qontract-reconcile/pull/3689/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL116-R144), [link](https://github.com/app-sre/qontract-reconcile/pull/3689/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL135-R160), [link](https://github.com/app-sre/qontract-reconcile/pull/3689/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL175-R204), [link](https://github.com/app-sre/qontract-reconcile/pull/3689/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL383-R414))
*  Modify `check_output` method of `TerraformClient` class to accept log argument, split it into lines, match provider logs with warnings or errors, and simplify error flag logic ([link](https://github.com/app-sre/qontract-reconcile/pull/3689/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL556-R603))
*  Add constants for provider log regex pattern and terraform log level in `terraform_client.py` ([link](https://github.com/app-sre/qontract-reconcile/pull/3689/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfR48-R49))
*  Update imports in `terraform_client.py` and `test_utils_terraform_client.py` to include new modules and classes ([link](https://github.com/app-sre/qontract-reconcile/pull/3689/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL3-R14), [link](https://github.com/app-sre/qontract-reconcile/pull/3689/files?diff=unified&w=0#diff-9ce65893311aef81c9bbccde0dcc4c1d44037684b52fe82331cb5376cd4c95daL2-R17))
*  Extract account name as constant in `test_utils_terraform_client.py` ([link](https://github.com/app-sre/qontract-reconcile/pull/3689/files?diff=unified&w=0#diff-9ce65893311aef81c9bbccde0dcc4c1d44037684b52fe82331cb5376cd4c95daL22-R37))
*  Add tests for `check_output` method of `TerraformClient` class in `test_utils_terraform_client.py` using different scenarios of return code, stdout, stderr, and log ([link](https://github.com/app-sre/qontract-reconcile/pull/3689/files?diff=unified&w=0#diff-9ce65893311aef81c9bbccde0dcc4c1d44037684b52fe82331cb5376cd4c95daR653-R883))

### CODE REVIEW

**Review by kwilczynski — APPROVED** [review]
@hemslo, looks good!

Nice feature! Will help us uncover issues.

File: reconcile/utils/terraform_client.py [hunk 1]
```diff
@@ -553,22 +553,26 @@ def check_output(
...
-            if return_code == 0:
-                for line in stderr:
+                if line.startswith("[WARN]"):
                     logging.warning(line_format.format(name, cmd, line))
-            else:
+                elif line.startswith("[ERROR]"):
+                    logging.error(line_format.format(name, cmd, line))
+                else:
+                    logging.debug(line_format.format(name, cmd, line))
+            if error_occured:
```
Comment (chassing) [inline reconcile/utils/terraform_client.py:570]: We're on Python 3.11 now. You may want to use structural pattern matching (`match ... case`) here?

Comment (hemslo) [reply reconcile/utils/terraform_client.py:570]: Great! Like this? Just checked all warn and error logs from https://github.com/hashicorp/terraform-provider-aws, there always be a space after `[WARN]` or `[ERROR]`.

```python
                match line.split():
                    case ["[WARN]", *_]:
                        logging.warning(line_format.format(name, cmd, line))
                    case ["[ERROR]", *_]:
                        logging.error(line_format.format(name, cmd, line))
                    case _:
                        logging.debug(line_format.format(name, cmd, line))
```

Comment (chassing) [reply reconcile/utils/terraform_client.py:570]: Yeah, exactly! Thx

**Review by chassing — APPROVED** [review]
❤️

**Comment by hemslo** [comment]
Local run showed that terraform logs are not in `stdout`, have to set `TF_LOG`, then it will be in `stderr`, https://developer.hashicorp.com/terraform/internals/debugging.

Another tricky part is even it's a warning log in aws provider, it will be part of info log of terraform like this
```
2023-07-20T15:49:09.681+1000 [INFO]  plugin.terraform-provider-aws_v3.76.0_x5: 2023/07/20 15:49:09 [WARN] DB Instance (xxx) not found, removing from state: timestamp=2023-07-20T15:49:09.673+1000
```

Any suggestions?

**Comment by chassing** [comment]
> Any suggestions?

Back to good old `if "[WARN]" in line: ...`?

**Comment by hemslo** [comment]
@chassing the tricky part is that provider log will be in `stderr`, we can't tell part the actual error log from terraform and the one from provider. If we conditionally check `[WARN]`, we may miss other valuable error info.

**Comment by chassing** [comment]
What about just relying completely on `TF_LOG`? I mean set `TF_LOG=WARN` and log everything? Or is `TF_LOG=JSON` better to parse?

**Comment by hemslo** [comment]
OK, here is another tricky part about `TF_LOG`, `[WARN]` from aws provider only appear in terraform `INFO` log, example in https://github.com/app-sre/qontract-reconcile/pull/3689#issuecomment-1643307141

**Comment by hemslo** [comment]
Thanks to @kwilczynski, new idea:
Use `TF_LOG_PATH` to let terraform write log to a seperate file instead of current stdout and stderr, so our current handling logic doesn't need to change.
Just add an additional handling for log file, use a regex to grab provider log out, include it in our own warning or error log.

**Review by kwilczynski — APPROVED** [review]
@hemslo, looks good!

Thank you!

**Pull request merged**